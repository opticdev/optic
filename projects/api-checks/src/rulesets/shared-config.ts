export enum RuleApplies {
  always = 'always',
  whenAdded = 'whenAdded',
  whenAddedOrChanged = 'whenAddedOrChanged',
}

export function ruleAppliesToLifeCycleKeyword(
  applies: RuleApplies
): 'requirement' | 'requirementOnChange' | 'added' {
  switch (applies) {
    case RuleApplies.always:
      return 'requirement';
    case RuleApplies.whenAddedOrChanged:
      return 'requirementOnChange';
    case RuleApplies.whenAdded:
      return 'added';
  }
}
