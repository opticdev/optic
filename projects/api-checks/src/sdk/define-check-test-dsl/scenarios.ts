import { OpenAPIV3 } from 'openapi-types';
const HttpMethods = OpenAPIV3.HttpMethods;
type ScenarioName = string;
export type BeforeAndAfter = [
  OpenAPIV3.Document,
  OpenAPIV3.Document,
  ScenarioName
];
export const copy = (obj: any) => JSON.parse(JSON.stringify(obj));
export type Editable<T> = (item: T) => T;

const defaultEmpty = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {},
});

const defaultWithoutSchema = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithResponseBodySchema = (
  schema: OpenAPIV3.SchemaObject | undefined
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
            content: {
              'application/json': {
                schema: schema,
              },
            },
          },
        },
      },
    },
  },
});

const defaultWithRequestParameters = (
  parameters: OpenAPIV3.ParameterObject[]
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        parameters,
        responses: {
          '200': {
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithResponseHeaders = (
  key: string,
  header: OpenAPIV3.HeaderObject,
  include: boolean
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            headers: include
              ? {
                  [key]: header,
                }
              : {},
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithRequestBodySchema = (
  schema: OpenAPIV3.SchemaObject | undefined
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      post: {
        requestBody: {
          content: {
            'application/json': {
              schema,
            },
          },
        },
        responses: {
          '201': {
            description: '',
          },
        },
      },
    },
  },
});

const baseParam = { in: 'query', name: 'exampleParam' };

export function scenario(name: string) {
  return {
    // responseSchema: {
    //   added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
    //     return [defaultWithoutSchema(), defaultWithResponseBodySchema(schema)];
    //   },
    //   changed: (
    //     beforeSchema: OpenAPIV3.SchemaObject,
    //     afterSchema: OpenAPIV3.SchemaObject
    //   ): BeforeAndAfter => {
    //     return [
    //       defaultWithResponseBodySchema(beforeSchema),
    //       defaultWithResponseBodySchema(afterSchema),
    //     ];
    //   },
    // },
    paths: {
      changed: (
        pathsBefore: OpenAPIV3.PathsObject,
        editPaths: Editable<OpenAPIV3.PathsObject>
      ): BeforeAndAfter => {
        return [
          { ...defaultEmpty(), paths: pathsBefore },
          { ...defaultEmpty(), paths: editPaths(copy(pathsBefore)) },
          name,
        ];
      },
    },
    operation: {
      requirement: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const spec: OpenAPIV3.Document = defaultEmpty();
        spec.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [spec, spec, name];
      },
      added: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        copied.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [defaultEmpty(), copied, name];
      },
      removed: (
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        delete copied.paths[pathPattern]?.[method];
        return [copied, defaultEmpty(), name];
      },
      changed: (
        operationBefore: OpenAPIV3.OperationObject,
        operationAfter: (
          before: OpenAPIV3.OperationObject
        ) => OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const before: OpenAPIV3.Document = defaultEmpty();
        before.paths = {
          [pathPattern]: {
            [method]: operationBefore,
          },
        };
        const after: OpenAPIV3.Document = defaultEmpty();
        after.paths = {
          [pathPattern]: {
            [method]: operationAfter(copy(operationBefore)),
          },
        };
        return [before, after, name];
      },
    },
    requestParameter: {
      added: (parameterToAdded: OpenAPIV3.ParameterObject): BeforeAndAfter => {
        return [
          defaultWithRequestParameters([]),
          defaultWithRequestParameters([parameterToAdded]),
          name,
        ];
      },
      changed: (
        parameterBefore: OpenAPIV3.ParameterObject,
        editParameter: Editable<OpenAPIV3.ParameterObject>
      ): BeforeAndAfter => {
        return [
          defaultWithRequestParameters([parameterBefore]),
          defaultWithRequestParameters([editParameter(copy(parameterBefore))]),
          name,
        ];
      },
    },
    responseHeader: {
      added: (key: string, header: OpenAPIV3.HeaderObject): BeforeAndAfter => {
        return [
          defaultWithResponseHeaders(key, header, false),
          defaultWithResponseHeaders(key, header, true),
          name,
        ];
      },
      changed: (
        key: string,
        headerBefore: OpenAPIV3.HeaderObject,
        editHeader: Editable<OpenAPIV3.HeaderObject>
      ): BeforeAndAfter => {
        return [
          defaultWithResponseHeaders(key, headerBefore, true),
          defaultWithResponseHeaders(key, editHeader(copy(headerBefore)), true),
          name,
        ];
      },
    },
    requestBodySchema: {
      changed: (
        schemaBefore: OpenAPIV3.SchemaObject,
        editSchema: Editable<OpenAPIV3.SchemaObject>
      ): BeforeAndAfter => {
        return [
          defaultWithRequestBodySchema(schemaBefore),
          defaultWithRequestBodySchema(editSchema(copy(schemaBefore))),
          name,
        ];
      },
      added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
        return [
          defaultWithRequestBodySchema(undefined),
          defaultWithRequestBodySchema(schema),
          name,
        ];
      },
    },
    responseBodySchema: {
      changed: (
        schemaBefore: OpenAPIV3.SchemaObject,
        editSchema: Editable<OpenAPIV3.SchemaObject>
      ): BeforeAndAfter => {
        return [
          defaultWithResponseBodySchema(schemaBefore),
          defaultWithResponseBodySchema(editSchema(copy(schemaBefore))),
          name,
        ];
      },
      added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
        return [
          defaultWithResponseBodySchema(undefined),
          defaultWithResponseBodySchema(schema),
          name,
        ];
      },
    },
  };
}
