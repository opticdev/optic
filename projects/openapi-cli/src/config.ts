export interface CliConfig {
  analytics: {
    segment: null | {
      key: string;
    };
  };
  errors: {
    sentry: null | {
      dsn: string;
    };
  };
}

export function readConfig(): CliConfig {
  return {
    analytics: {
      segment: process.env.OPTIC_OPENCLI_SEGMENT_KEY
        ? {
            key: process.env.OPTIC_OPENCLI_SEGMENT_KEY,
          }
        : null,
    },
    errors: {
      sentry: process.env.OPTIC_OPENCLI_SENTRY_DSN
        ? {
            dsn: process.env.OPTIC_OPENCLI_SENTRY_DSN,
          }
        : null,
    },
  };
}
