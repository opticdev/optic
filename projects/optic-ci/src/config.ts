import path from 'path';
import fs from 'fs';
import { Rule, Ruleset } from '@useoptic/rulesets-base';
import { RulesetDefinition as SpectralRulesetDefinition } from '@stoplight/spectral-core';

import { OPTIC_CONFIG_PATH } from './constants';
/**
 * Expected format
 * module.exports = {
 *  token: process.env.OPTIC_TOKEN,
 *  gitProvider: {
 *    token: process.env.GITHUB_TOKEN,
 *  },
 * rules: [],
 * generateContext: (details: {
 *  fileName: string
 * }) => ({})
 */

export type OpticConfiguration = {
  token?: string;
  gitProvider?: {
    token: string;
  };
  rules?: (Ruleset | Rule)[];
  generateContext?: (details: { fileName: string }) => Object;
  spectralConfig?: Extract<
    SpectralRulesetDefinition,
    { extends: any; rules: any }
  >['rules'];
};

const findOpticConfigPath = (dir: string): string | null => {
  while (true) {
    const opticConfigPath = path.join(dir, OPTIC_CONFIG_PATH);

    if (fs.existsSync(opticConfigPath)) {
      return opticConfigPath;
    }

    const nextDir = path.dirname(dir);
    if (nextDir === dir) {
      break;
    }
    dir = nextDir;
  }

  return null;
};

export const readConfig = async (): Promise<OpticConfiguration> => {
  const opticConfigPath = findOpticConfigPath(process.cwd());
  if (opticConfigPath) {
    // TODO add validation to configuration loading
    // TODO add in fallbacks / defaults here
    return import(opticConfigPath);
  } else {
    return {};
  }
};
