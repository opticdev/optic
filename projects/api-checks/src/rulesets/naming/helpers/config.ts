import { RuleApplies } from '../../shared-config';

export enum NameMustBe {
  none = 'none',
  snakeCase = 'snake_case',
  camelCase = 'camelCase',
  paramCase = 'param-case',
  pascalCase = 'PascalCase',
}

export interface NameRule {
  rule: NameMustBe;
  applies: RuleApplies;
}

export interface NamingChecksConfig {
  requestHeaders?: NameRule;
  queryParameters?: NameRule;
  requestProperties?: NameRule;
  responseProperties?: NameRule;
  responseHeaders?: NameRule;
}
