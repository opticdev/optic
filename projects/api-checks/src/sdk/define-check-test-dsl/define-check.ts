import { BeforeAndAfter } from './scenarios';
import { ApiCheckService } from '../api-check-service';
import { ApiChangeDsl } from '../api-change-dsl';
import { factsToChangelog } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export interface IApiCheckDefinition<CheckConfig> {
  name: string;
  implementation?: (apiChangeDsl: ApiChangeDsl, config: CheckConfig) => void;
  description?: string;
  validExamples: BeforeAndAfter[];
  invalidExamples: BeforeAndAfter[];
}

export class ApiCheckImpl<CheckConfig> {
  public check: IApiCheckDefinition<CheckConfig>;

  constructor(name: string) {
    this.check = {
      name,
      validExamples: [],
      invalidExamples: [],
    };
  }

  description(description: string) {
    this.check.description = description;
    return this;
  }

  implementation(
    implementation: (
      apiChangeDsl: ApiChangeDsl,
      config?: CheckConfig
    ) => void | Promise<void>
  ) {
    this.check.implementation = implementation;
    return this;
  }

  runner() {
    return (dsl: ApiChangeDsl) =>
      this.check.implementation!(dsl, undefined as any);
  }

  runnerWithConfig(config: CheckConfig) {
    return (dsl: ApiChangeDsl) => this.check.implementation!(dsl, config);
  }

  passingExample(beforeAndAfter: BeforeAndAfter, config?: CheckConfig) {
    this.check.validExamples = [...this.check.validExamples, beforeAndAfter];

    if (!this.check.implementation) test.skip('', () => {});

    if (
      typeof jest !== 'undefined' &&
      this.check.implementation &&
      process.env['RUN_CHECK_JEST_TESTS']
    ) {
      // @ts-ignore
      const { test, expect } = global;

      const testName = `passing case ${this.check.validExamples.length}: ${beforeAndAfter[2]}`;

      test(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter, config);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeTruthy();
      });
    }

    return this;
  }

  failingExample(beforeAndAfter: BeforeAndAfter, config?: CheckConfig) {
    this.check.invalidExamples = [
      ...this.check.invalidExamples,
      beforeAndAfter,
    ];

    if (!this.check.implementation) test.skip('', () => {});

    if (
      typeof jest !== 'undefined' &&
      this.check.implementation &&
      process.env['RUN_CHECK_JEST_TESTS']
    ) {
      // @ts-ignore
      const { it, expect } = global;

      const testName = `failing case ${this.check.invalidExamples.length}: ${beforeAndAfter[2]}`;
      it(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter, config);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeFalsy();
      });
    }

    return this;
  }

  private async testExample(
    beforeAndAfter: BeforeAndAfter,
    config?: CheckConfig
  ) {
    invariant(this.check.implementation);
    const service = new ApiCheckService().useRulesFrom((dsl) =>
      this.check.implementation!(dsl, config as any)
    );

    const [before, after] = beforeAndAfter;
    const { currentFacts, nextFacts } = service.generateFacts(before, after);
    const runResults = await service.runRulesWithFacts({
      context: {},
      nextFacts,
      currentFacts,
      changelog: factsToChangelog(currentFacts, nextFacts),
      nextJsonLike: after,
      currentJsonLike: before,
    });

    return {
      passed: !runResults.some((i) => !i.passed),
      results: runResults,
    };
  }
}

export function check<CheckConfig = undefined>(name: string) {
  return new ApiCheckImpl<CheckConfig>(name);
}
