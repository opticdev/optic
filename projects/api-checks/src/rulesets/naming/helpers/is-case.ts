import { NameMustBe } from './config';

function regexForRule(nameRule: NameMustBe) {
  switch (nameRule) {
    case NameMustBe.camelCase:
      return /^[a-z][a-z0-9]*(?:[A-Z0-9][a-z0-9]+)*$/;
    case NameMustBe.capitalParamCase:
      return /^[A-Z0-9][a-z0-9]*(-[A-Z0-9][a-z0-9]*)*$/;
    case NameMustBe.paramCase:
      return /^[a-z0-9]+(-[a-z0-9]+)*$/;
    case NameMustBe.pascalCase:
      return /^[A-Z][a-z0-9]+(?:[A-Z0-9][a-z0-9]+)*$/;
    case NameMustBe.snakeCase:
      return /^[a-z0-9]+(?:_[a-z0-9]+)*$/;
    default:
      return /(.*?)/;
  }
}

export function isCase(example: string, nameRule: NameMustBe): boolean {
  return regexForRule(nameRule).test(example);
}
