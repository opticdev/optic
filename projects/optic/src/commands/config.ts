import { Command } from 'commander';
import { OpticCliConfig, RenderTemplate } from '../config';
import * as yaml from 'yaml';

export const registerConfigCommand = (
  cli: Command,
  cliConfig: OpticCliConfig
) => {
  const config = cli.command('config');
  config
    .helpOption('-h, --help', 'Display help for the command')
    .addHelpCommand(false)
    .configureHelp({ commandUsage: () => `\n  optic config <command>` })
    .description('Commands to manage the Optic configuration');

  config
    .command('show')
    .description('Display the rendered Optic configuration from optic.yml')
    .configureHelp({ commandUsage: () => `\n  optic config show` })
    .action(async () => {
      console.log(
        // rendering the template a second time isn't awesome (it first happens when the CLI is initialized), but an OpticCliConfig contains more than just whats in optic.yml and doesn't have a simple way to convert to yaml anyway.
        yaml.stringify(await RenderTemplate(cliConfig.configPath!), null, 2)
      );
    });
};
