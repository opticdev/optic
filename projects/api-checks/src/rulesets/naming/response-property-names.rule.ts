import { check } from '../../sdk/define-check-test-dsl/define-check';
import { scenario } from '../../sdk/define-check-test-dsl/scenarios';
import { NameMustBe, NameRule } from './helpers/config';
import { RuleApplies, ruleAppliesToLifeCycleKeyword } from '../shared-config';
import { isCase } from './helpers/is-case';
import { expect } from 'chai';

export default check<NameRule>('response body property naming')
  .implementation(({ responses }, config) => {
    const applies = config?.applies || RuleApplies.always;
    const casing = config?.rule || NameMustBe.none;

    const { expect } = require('chai');
    responses.bodyProperties[ruleAppliesToLifeCycleKeyword(applies)].must(
      `have ${casing.toString()} name`,
      (current) => {
        if (!isCase(current.key, casing))
          expect.fail(`${current.key} is not ${casing.toString()}`);
      }
    );
  })
  .failingExample(
    scenario(
      'adding a camelCase when pascalCase is required'
    ).responseBodySchema.added({
      type: 'object',
      properties: {
        examplePropertyName: { type: 'string' },
      },
      required: [],
    }),
    { rule: NameMustBe.pascalCase, applies: RuleApplies.always }
  )
  .passingExample(
    scenario(
      'adding a camelCase when camelCase is required'
    ).responseBodySchema.added({
      type: 'object',
      properties: {
        camelCase: { type: 'string' },
      },
      required: [],
    }),
    { rule: NameMustBe.camelCase, applies: RuleApplies.always }
  )
  .passingExample(
    scenario('wrong case ok if it is already there').responseBodySchema.changed(
      {
        type: 'object',
        properties: {
          examplePropertyName: { type: 'string' },
        },
        required: [],
      },
      (same) => same
    ),
    { rule: NameMustBe.pascalCase, applies: RuleApplies.whenAdded }
  );
