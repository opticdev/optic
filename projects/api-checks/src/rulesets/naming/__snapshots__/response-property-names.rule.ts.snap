// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failing case 1: adding a camelCase when pascalCase is required 1`] = `
Object {
  "passed": false,
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "examplePropertyName",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "examplePropertyName",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/examplePropertyName",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "examplePropertyName",
          "required": false,
        },
      },
      "condition": "have PascalCase name",
      "docsLink": undefined,
      "effectiveOnDate": undefined,
      "error": "examplePropertyName is not PascalCase",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for field: response body 'examplePropertyName'",
    },
  ],
}
`;

exports[`passing case 1: adding a camelCase when camelCase is required 1`] = `
Object {
  "passed": true,
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "body": Object {
                "contentType": "application/json",
              },
              "statusCode": "200",
            },
            "jsonSchemaTrail": Array [
              "camelCase",
            ],
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "application/json",
            "camelCase",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/content/application~1json/schema/properties/camelCase",
          "kind": "field",
        },
        "value": Object {
          "flatSchema": Object {
            "type": "string",
          },
          "key": "camelCase",
          "required": false,
        },
      },
      "condition": "have camelCase name",
      "docsLink": undefined,
      "effectiveOnDate": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for field: response body 'camelCase'",
    },
  ],
}
`;

exports[`passing case 2: wrong case ok if it is already there 1`] = `
Object {
  "passed": true,
  "results": Array [],
}
`;
